### ISSUE TITLE
feat(finetune): add DPO — Direct Preference Optimization for alignment

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#449]
## Summary

Add Direct Preference Optimization (DPO) as a fine-tuning method. DPO is the
alignment stage in the adaptation pipeline — it teaches the model to prefer
better outputs over worse ones using human preference data. Replaces RLHF with
a simpler, more stable objective.

## CLI

```bash
apr finetune model.apr --method dpo \
  --data preferences.jsonl \
  --dpo-beta 0.1 \
  --epochs 3 \
  -o aligned-model.apr
```

## Data Format (JSONL)

```json
{"prompt": "Explain quantum computing", "chosen": "preferred response...", "rejected": "worse response..."}
```

## Parameters

- `--dpo-beta`: KL penalty weight (default: 0.1). Higher = more conservative.
- `--dpo-label-smoothing`: Label smoothing (default: 0.0).
- `--ref-model`: Reference model path (default: use input model as reference).

## DPO Loss

```
L_DPO = -log σ(β * (log π(chosen)/π_ref(chosen) - log π(rejected)/π_ref(rejected)))
```

## Implementation

1. Add `dpo` variant to FinetuneMethod enum
2. DPO data loader: parse prompt/chosen/rejected triples
3. Reference model loading (freeze, use for log-prob computation)
4. DPO loss implementation
5. Support LoRA (train adapter with DPO objective)
6. Checkpoint compatible with subsequent RLVR stage

## Arcee Reference

Arcee's SLM adaptation system uses DPO as the third stage after CPT and SFT.
They offer DPO training via both GUI and arcee-py SDK.

## Specification

See `docs/specifications/components/finetune.md` §5.2.

### Contract: `dpo-alignment-v1.yaml`

```yaml
metadata:
  description: "Direct Preference Optimization — alignment via preference pairs"
  references:
    - "Rafailov et al. (2023) Direct Preference Optimization"
  depends_on:
    - "cross-entropy-kernel-v1"
    - "adamw-kernel-v1"
    - "lora-algebra-v1"

equations:
  dpo_loss:
    formula: "L_DPO = -log σ(β * (log π(y_w|x)/π_ref(y_w|x) - log π(y_l|x)/π_ref(y_l|x)))"
    domain: "β > 0, y_w = chosen, y_l = rejected"
    invariants:
      - "L_DPO >= 0 (log-sigmoid is non-positive, negated)"
      - "As π(y_w)/π_ref(y_w) increases, loss decreases"
      - "β = 0 → L = log(2) (uniform prior)"

  implicit_reward:
    formula: "r(y|x) = β * log(π(y|x) / π_ref(y|x))"
    invariants:
      - "r(y_w|x) > r(y_l|x) when training converges"
      - "Reference model is frozen throughout training"

proof_obligations:
  - type: bound
    property: "DPO loss non-negativity"
    formal: "L_DPO >= 0 for all inputs"
  - type: monotonicity
    property: "Reward margin increase"
    formal: "r(y_w) - r(y_l) increases during training"
  - type: invariant
    property: "Reference model frozen"
    formal: "π_ref weights unchanged throughout DPO training"
  - type: invariant
    property: "β positivity"
    formal: "β > 0 enforced at construction"

falsification_tests:
  - id: FALSIFY-DPO-001
    rule: "Loss non-negative"
    prediction: "DPO loss >= 0 for random preference pairs"
    if_fails: "Log-sigmoid computation or negation bug"
  - id: FALSIFY-DPO-002
    rule: "Reference frozen"
    prediction: "π_ref weights identical before and after training"
    if_fails: "Reference model accidentally updated by optimizer"
  - id: FALSIFY-DPO-003
    rule: "β=0 degeneracy"
    prediction: "L_DPO = log(2) when β=0"
    if_fails: "β scaling not correctly applied"
  - id: FALSIFY-DPO-004
    rule: "Preference direction"
    prediction: "swapping chosen/rejected inverts the loss gradient direction"
    if_fails: "Chosen/rejected labels confused in loss computation"

kani_harnesses:
  - id: KANI-DPO-001
    obligation: "DPO loss non-negativity"
    property: "L_DPO >= 0 for bounded logprob inputs"
    bound: 4
    strategy: bounded_int
```

### Binding Requirements

Add to `contracts/entrenar/binding.yaml`:
```yaml
  - contract: dpo-alignment-v1.yaml
    equation: dpo_loss
    module_path: "entrenar::finetune::dpo"
    function: dpo_loss
    status: not_implemented

  - contract: dpo-alignment-v1.yaml
    equation: implicit_reward
    module_path: "entrenar::finetune::dpo"
    function: compute_reward
    status: not_implemented
```

### Verification Targets
- L2: 4 falsification tests
- L3: Kani harness for loss non-negativity
- L4: Lean proof for β=0 degeneracy (log σ(0) = -log(2))
